File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    global maxDiff, answer
    answer = [-1]
    tmp = [0]*11
    maxDiff = 0
    
    def find_gap(score):
        
        apich, rian = 0, 0
        for i in range(len(score)):
            if score[i] > 0 or info[i] > 0:
                if score[i] > info[i]:
                    rian += (10 - i)
                else:
                    apich += (10-i)
        return (rian>apich, abs(rian- apich))
    
    
    def dfs(L, cnt):
        global maxDiff, answer
        if L == 11 or cnt ==0:
            winner, gap = find_gap(tmp)
            
            if winner:
                if cnt >= 0:
                    tmp[10] = cnt
                if gap > maxDiff:
                    maxDiff = gap
                    answer = tmp.copy()

                    
                elif gap == maxDiff:
                    for i in range(len(tmp)):
                        if answer[i] > 0:
                            a = i
                        if tmp[i] > 0 :
                            b = i
                    if b > a :
                        answer = tmp.copy()
                        
            return
            

        if cnt >info[L]:
            tmp[L] = info[L]+1
            dfs(L+1, cnt - (info[L]+1))
            tmp[L] = 0
        
        dfs(L+1, cnt)
    

    dfs(0,n)
    return answer

File: 프로그래머스/2/92342. 양궁대회/양궁대회.java
import java.util.*;
class Solution {
    static int[] rian;
    static int n;
    static int[] info;
    static int[] answer = {-1};
    static int max = Integer.MIN_VALUE;
    public int[] solution(int n, int[] info) {
        this.n = n;
        this.info = info;
        rian = new int[11];
        
        dfs(0, 0, rian);
        if(max == -1){
            int[] res = {-1};
            return res;
        }
        return answer;
    }
    public static void dfs(int L, int start, int[] rian){
        if(L==n){
            int diff = getScore(rian);
            if(max<diff){
                max = diff;
                answer = rian.clone();
            }else if(max == diff){
                int cnt = 0;
                for(int i=10;i>-1;i--){
                    if(answer[i]<rian[i]){
                        answer = rian.clone();
                        break;
                    }
                    if(answer[i]>rian[i]){
                        break;
                    }
                        
                }
            }
            return;
            
        }
        
        for(int i=start; i<11;i++){
            if(info[i] < rian[i]) continue;
            rian[i] ++;
            dfs(L+1,i,rian);
            rian[i]--;
        }
    }
    public static int getScore(int[] arr){
        int cnt = 0;
        int a =0;
        int r = 0;
        for(int i=0; i<11;i++){
            if(info[i] ==0 && arr[i] == 0) continue;
            if(info[i]<arr[i]) r+=10-i;
            else a+=10-i;
        }
        cnt = r-a;
        if(cnt<=0){
            return -1;
        }
        return cnt;
    }
}